CVLByExample/Ecrecover/ecrecover.spec

```diff
@@ -0,0 +1,278 @@
+    assert ecrecover(msgHashA, v, r, s) != 0 => ecrecover(msgHashB, v, r, s) == 0;
+}
+
+rule dependencyOnS() {
+    ecrecoverAxioms();
+    bytes32 msgHash;  uint8 v; bytes32 r; bytes32 s1; bytes32 s2; 
+    require s1 != s2;  
+    assert ecrecover(msgHash, v, r, s1) != 0 => ecrecover(msgHash, v, r, s2) == 0;
+    //!= ecrecover(msgHash, v, r, s1) ;
+}
+
+rule dependencyOnR() {
+    ecrecoverAxioms();
+    bytes32 msgHash; uint8 v; bytes32 r1; bytes32 r2; bytes32 s; 
+    require r1 != r2;  
+    assert ecrecover(msgHash, v, r1, s) != 0 => ecrecover(msgHash, v, r2, s) == 0 ;
+}
+
+// Rules for a function that checks signature 
+
+rule singleVerifier () 
+{
+    ecrecoverAxioms();
+    bytes32 msgHash; uint8 v; bytes32 r; bytes32 s; 
+    address addr1; 
+    address addr2; 
+    require addr1 != addr2 && addr1!=0 && addr2!=0 ;
+    assert isSigned(addr1, msgHash, v, r, s ) => !isSigned(addr2, msgHash, v, r, s ) ;
+
+}
+
+
+rule ownerSignatureIsUnique () {
+    ecrecoverAxioms();
+    bytes32 msgHashA; bytes32 msgHashB;
+    uint8 v; bytes32 r; bytes32 s; 
+    address addr; 
+    require msgHashA != msgHashB; 
+    require addr != 0;
+    assert isSigned(addr, msgHashA, v, r, s ) => !isSigned(addr, msgHashB, v, r, s );
+}
+
+rule hashIsUnique(
+    address ownerA,
+    uint256 myParamA,
+    uint256 deadlineA,
+    address ownerB,
+    uint256 myParamB,
+    uint256 deadlineB
+) {
+    bytes32 hashA =  getHash(ownerA, myParamA, deadlineA);
+    bytes32 hashB = getHash(ownerB, myParamB, deadlineB);
+    assert hashA == hashB => (  ownerA == ownerB && 
+                                myParamA == myParamB &&
+                                deadlineA == deadlineB ); 
+}
+
+
+/*** 
+   # High level property : there is only single owner that can be used
+    A rule which proves that for a given set of parameters only a single owner can execute .
+    This property is implemented as a relational property - it compares two different executions on the same state.
+*/
+rule onlySingleUserCanExecute(uint8 v,
+    bytes32 r,
+    bytes32 s,
+    address alice,
+    address bob,
+    uint256 myParam,
+    uint256 deadline) {
+
+    env e1;
+    env e2;
+    ecrecoverAxioms();
+    //save the current state
+    storage init = lastStorage; 
+    //execute and assume succeeded
+    address temp = ecrecover(getHash(bob, myParam, deadline), v, r, s);
+    executeMyFunctionFromSignature(e1, v, r, s, alice, myParam, deadline);
+    // compare another execution on the same state, look at reverting paths
+    executeMyFunctionFromSignature@withrevert(e2, v, r, s, bob, myParam, deadline) at init;
+    bool success = !lastReverted;
+    //if it succeeded it must be the same owner, or an unrealistic hash collision
+    assert success => ( alice==bob ) ; 
+}
+
+
+/*** 
+   # High level property : there is only single owner that can be used
+    A rule which proves that for a given set of parameters only a single owner can execute .
```

Comment:
fix comment Owner must be the signer of the hash

Reply:
done